### PR TITLE
Add loading mask to d2-ui-components

### PR DESCRIPTION
**Pull request by SferaDev**
### :memo: Implementation

- Add a Loading Mask Provider

### :art: Screenshots

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -121,6 +121,44 @@ MyComponent.propTypes = {
...
+To use it, create a HOC with `withLoading`, add `loading` to your propTypes, and show the mask using the functions `props.loading.show()`.
+
+```
+import { withLoading } from "d2-ui-components";
+import PropTypes from "prop-types";
+
+const MyComponent = ({name, loading}) => (
+    <div>
+        <a onClick={() => loading.show()}>Show loading mask</a>
+        <a onClick={() => loading.show(false)}>Hide loading mask</a>
```
Comment (tokland) [inline README.md:166]: `loading.show(boolean)` is very useful, but can we have also `loading.hide()` for the simple case?

File: src/loading/LoadingConsumer.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    },
+    divider: {
+        margin: "40px"
+    }
+});
+
+const LoadingConsumer = props => {
+    const { classes } = props;
+
+    const showMessage = () => {
```
Comment (tokland) [inline src/loading/LoadingConsumer.js]: Unused function.

File: src/loading/LoadingConsumer.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                            />
+                            <Divider
+                                className={classes.divider}
+                                variant="middle"
+                                hidden={!message || message.trim().length === 0}
+                            />
+                            <Typography
+                                className={classes.message}
+                                variant="h6"
+                                hidden={!message || message.trim().length === 0}
```
Comment (tokland) [inline src/loading/LoadingConsumer.js]: Used two times, maybe move it a to local variable `hidden`?

File: src/loading/LoadingConsumer.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import React from "react";
```
Comment (tokland) [inline src/loading/LoadingConsumer.js]: Should be a jsx file. I see why you have a js file, I had it wrong for the snack: `src/snackbar/SnackbarConsumer.js`, could be changed also there.

File: src/loading/LoadingConsumer.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import PropTypes from "prop-types";
+
+import { withStyles } from "@material-ui/core/styles";
+import LoadingContext from "./context";
+import CircularProgress from "@material-ui/core/CircularProgress";
+import Typography from "@material-ui/core/Typography";
+import Divider from "@material-ui/core/Divider";
+
+const styles = theme => ({
```
Comment (tokland) [inline src/loading/LoadingConsumer.js]: eslint does not like non-used arguments, `_theme` should do the trick.

File: src/loading/LoadingProvider.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import React, { Component } from "react";
```
Comment (tokland) [inline src/loading/LoadingProvider.js]: Same, jsx file.

File: src/loading/LoadingProvider.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import LoadingContext from "./context";
+import LoadingConsumer from "./LoadingConsumer";
+
+export default class SnackbarProvider extends Component {
```
Comment (tokland) [inline src/loading/LoadingProvider.js]: LoadingProvider

**Review by tokland — COMMENTED** [review]
The code (and the snapshot, nice!) looks good to me, just some minor things.

Btw, Is there a branch in metadata-sync that uses it, so we can check how it works? Sometime we should create a simple dirty app with all the components on one page so we can them easily.

File: README.md (hunk 1 above)
Comment (SferaDev) [reply README.md:166]: Sounds good, will add.

File: src/loading/LoadingConsumer.js (hunk 4 above)
Comment (SferaDev) [reply src/loading/LoadingConsumer.js]: Yeah I just started the branch off Snackbar with shallow copy. Will change them in this PR now

File: src/loading/LoadingConsumer.js (hunk 3 above)
Comment (SferaDev) [reply src/loading/LoadingConsumer.js]: Okay I'll move to render variable

Comment (tokland) [reply src/loading/LoadingConsumer.js]: Now that we are at it, isn't equivalent to: `!message || !message.trim()`?

**Comment by SferaDev** [comment]
> Btw, Is there a branch in metadata-sync that uses it, so we can check how it works? Sometime we should create a simple dirty app with all the components on one page so we can them easily.

Not yet but we should have it before Friday. Anyway I think the best approach is creating a [StoryBook](https://github.com/storybooks/storybook). It's really cool and now it's like the state-of-the-art solution for ReactJS examples.

**Comment by tokland** [comment]
> Anyway I think the best approach is creating a [StoryBook](https://github.com/storybooks/storybook). 

Wow, it really looks nice. I think it still makes sense to  have an `example` CRA in the repo itself, but it looks good. Maybe you can show us on the next team meeting.

**Review by tokland — APPROVED** [review]
I have not tested functionally, I'll wait for CRA example (or metadata-sync test branch), but the code looks very good to me.

**Review by adrianq — APPROVED** [review]
I tested all the features in the multiple metadata-sync PR. Awesome work guys!

**Pull request merged**